### PR TITLE
sys_fs & VFS: Misc bug fixes

### PR DESCRIPTION
**Pull request by brian218**
1. Do not mount dev_usbXXX if its path is not properly specified since this can cause confusion to some games that detect and determine which USB flash drive to use.
2. Allow dev_usbXXX to be mounted, unmounted, and formatted(newfs) by the corresponding sys_fs syscalls as some games do format the attached USB flash drives in some cases.
3. Improve the permission check for the three syscalls mentioned above: PS3 root permission is required in order to perform actions on devices other than dev_usbXXX.
4. Add read-only flag check to sys_fs_newfs() to prevent devices with this flag from being formatted.
5. Implement CELL_FS_PATH path-as-device support.

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 1]
```diff
@@ -336,16 +336,16 @@ lv2_fs_object::lv2_fs_object(utils::serial& ar, bool)
 u64 lv2_file::op_read(const fs::file& file, vm::ptr<void> buf, u64 size)
 {
 	// Copy data from intermediate buffer (avoid passing vm pointer to a native API)
-	uchar local_buf[65536];
+	std::vector<uchar> local_buf(size);
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:339]: Simply allocate min(65536, size), this was a workaround for when passing some huge size such as above 4GB.

Comment (Megamouse) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:339]: Isn't this completely busted anyway? It seems to memcpy the whole size into the 65536 buffer...

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:339]: There was an std::min there.

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 2]
```diff
@@ -341,16 +336,16 @@ lv2_fs_object::lv2_fs_object(utils::serial& ar, bool)
...
-	uchar local_buf[65536];
+	std::vector<uchar> local_buf(std::min<u64>(size, 65536));
 
 	u64 result = 0;
 
 	while (result < size)
 	{
-		const u64 block = std::min<u64>(size - result, sizeof(local_buf));
-		const u64 nread = file.read(+local_buf, block);
+		const u64 block = size - result;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:345]: this obviously still has to be 
```suggestion
		const u64 block = std::min<u64>(size - result, local_buf.size());
```

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 3]
```diff
@@ -365,15 +360,15 @@ u64 lv2_file::op_read(const fs::file& file, vm::ptr<void> buf, u64 size)
...
+	std::vector<uchar> local_buf(std::min<u64>(size, 65536));
 
 	u64 result = 0;
 
 	while (result < size)
 	{
-		const u64 block = std::min<u64>(size - result, sizeof(local_buf));
-		std::memcpy(local_buf, static_cast<const uchar*>(buf.get_ptr()) + result, block);
-		const u64 nwrite = file.write(+local_buf, block);
+		const u64 block = size - result;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:369]:
```suggestion
		const u64 block = std::min<u64>(size - result, local_buf.size());
```

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp (hunk 2 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp (hunk 3 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Fixed.

**Review by elad335 — COMMENTED** [review]
I rather not adding new lv2_obj::sleep calls, I know it's like other syscalls, but its placemrnt is questinable in many places.

**Comment by brian218** [comment]
> I rather not adding new lv2_obj::sleep calls, I know it's like other syscalls, but its placemrnt is questinable in many places.

How about `ppu.state += cpu_flag::wait`?

**Comment by elad335** [comment]
That statement is okay.

**Comment by brian218** [comment]
Reverted.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 4]
```diff
@@ -207,6 +207,7 @@ lv2_fs_mount_point* lv2_fs_object::get_mp(std::string_view filename, std::string
 {
 	auto result = &g_mp_sys_no_device; // Default fallback
 	const auto mp_name = get_device_path(filename);
+	std::string_view path_fix;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:210]:
```suggestion
	std::string path_fix;
```

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Done.

Comment (Megamouse) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Because a stringview is not supposed to be

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 5]
```diff
@@ -286,36 +299,24 @@ std::string lv2_fs_object::device_name_to_path(std::string_view device_name)
...
 
 	if (local_path.empty())
-	{
-		if (no_error)
-		{
-			return true;
-		}
-		else
-		{
-			sys_fs.error("\"%s\" is not mounted!", vpath);
```
Comment (Darkhost1999) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:315]: I don't see where this error message was moved to. Was it removed? If so why?
I keep noticing dev_hdd1 isn't mounted messages when I attempt to boot a game in xmb for testing purposes.
Therefore I think removing this message unless it is incorrect will fail to provide information that may be useful in developing the emulator in the future.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 6]
```diff
@@ -286,36 +299,24 @@ std::string lv2_fs_object::device_name_to_path(std::string_view device_name)
 					return fmt::format("%s%03d", mp->root, i);
 				}
 			}
+			break;
 		}
-		else
+		else if (mp->device == device_name || alias_check(mp))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:304]: Else is obsolete if you break in if

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 7]
```diff
@@ -329,7 +330,10 @@ bool lv2_fs_object::vfs_unmount(std::string_view vpath, bool no_error)
 		}
 	}
 
-	return vfs::unmount(vpath);
+	if (!vfs::unmount(vpath))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:333]: Unnecessary change

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp (hunk 5 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:315]: I thought the built-in error report mechanism was enough for this purpose.
See line 3236~3237:
`
	if (!lv2_fs_object::vfs_unmount(vpath))
		return {CELL_ENOTMOUNTED, vpath};
`

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp (hunk 6 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Done,

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp (hunk 7 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp (hunk 5 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:315]: For instance: `SYS: 'sys_fs_unmount' failed with 0x8001003a : CELL_ENOTMOUNTED, /dev_hdd1 [1]`
The above message itself is self-explanatory enough.

**Comment by brian218** [comment]
Excuse me, I wonder what else can still be keeping this PR from going ahead.

**Review by elad335 — APPROVED** [review]
